### PR TITLE
V2 HdrHistogram Log format (v1.3)

### PR DESCRIPTION
**Pull request by filipecosta90**
This is a PR to kick off V2 HdrHistogram Log format support on the go version. 

I've followed the wiki available at https://github.com/HdrHistogram/HdrHistogram.NET/wiki/V2-Log-Encoding and revised the python and java implementations. 

IMHO, we need:
- [x] Deeper testing ( included white box as well  specifically for zigzag and other critical internal parts )
- [x] Parse `test/jHiccup-2.0.7S.logV2.hlog` and `/test/tagged-Log.logV2.hlog`. 
- [x] Examples of using the log readers and writers:
  - histogram.Histogram.RecordValue
  - histogram.New
  - histogram.NewHistogramLogReader
  - histogram.NewHistogramLogWriter 
- [x] Better error handing
- [x] linter pass to double check all errors are handled.

### CODE REVIEW

**Comment by giltene** [comment]
(This is based on cursory reading). I think the zigzag decide and encode may be wrong, V2 uses a ZigZag-encoded LEB128-64b9B-variant that differs from the original LEB128 used in e.g. google protocol buffers. The difference us subtle: an intentional lack of a termination but in the last byte, which reduces the maximum encoding length of a 64 bit value from 10 bytes to 9. The current code looks like it may be for the more generic 64b10 case. The loop termination in the current code may fail to terminate on some values (high bit of highest byte set).
In addition, the reason for the “ugly” deeply nested ifs (rather than using a loop with termination conditions) in the Java implementation is speed: the whole point in these encodings is that the vast majority of values are encoded in a small number (usually 1) of bytes, the non-loop forms tend to compile into faster code for these common cases (on most optimizing compilers) with real benefits in e.g. deciding speed,

File: hdr_encoding.go [hunk 1]
```diff
@@ -226,39 +226,125 @@ func (rh *Histogram) fillBufferFromCountsArray() (buffer []byte, err error) {
...
-		n++
-		value |= (b & 0x7f) << shift
-		if (b & 0x80) == 0 {
-			break
+	if len(buf) < 1 {
+		return 0, 0
+	}
+	var value = uint64(buf[0]) & 0x7f
+	n = 1
+	if len(buf) > 1 && (buf[0]&0x80) != 0 {
```
Comment (giltene) [inline hdr_encoding.go]: Error handling is probably needed here. When additional bytes are indicated (by buf[i]&0x80) but those bytes are not available in the buffer, returning a value without an error indication will lead to a silent reading of corrupt values. Truncated compressed histogram strings are a real thing (through various communication or buffer overflow situations), so these errors to happen and should be made visible to the caller.

Comment (filipecosta90) [reply hdr_encoding.go]: agree :+1: , will change the signature from `zig_zag_decode_i64(buf []byte) (signedValue int64, n int)` to `zig_zag_decode_i64(buf []byte) (signedValue int64, n int, err error)`

Comment (filipecosta90) [reply hdr_encoding.go]: fixed in the latest commits @giltene . Also added spefic test for this at https://github.com/HdrHistogram/hdrhistogram-go/pull/35/files#diff-39e1e370a64ae2eadefea4bdd36d4080R26

File: zigzag.go [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+					if (buf[4] & 0x80) != 0 {
+						if buflen < 6 {
+							err = fmt.Errorf(truncatedErrStr, 6, buflen)
+							return
+						}
+						value |= uint64(buf[5]) & 0x7f << 35
+						n = 6
+						if (buf[5] & 0x80) != 0 {
+							if buflen < 7 {
+								err = fmt.Errorf(truncatedErrStr, 6, buflen)
```
Comment (giltene) [inline zigzag.go:52]: looks like some copy pasting of the fmt.Errorf lines, and the numbers after 6 aren't right (should be 7,8,9 I think)

**Comment by filipecosta90** [comment]
@giltene @ahothan I've revised all change requests. I believe we're good for a new review :)

**Pull request merged**